File: frameworks/mlr3automl/exec.R
library(mlr3)
library(mlr3learners)
library(mlr3extralearners)
library(mlr3tuning)
library(mlr3pipelines)
library(mlr3automl)
library(mlr3oml)

run <- function(train_file, test_file, target.index, type, output_predictions_file, cores, time.budget, meta_results_file, seed, name) {
  future::plan(future::multicore)
  start_time <- Sys.time()
  set.seed(seed)
  
  train <- mlr3oml::read_arff(train_file)
  colnames(train) <- make.names(colnames(train))
  target <- colnames(train)[target.index]

  test <- mlr3oml::read_arff(test_file)
  colnames(test) <- make.names(colnames(test))
  
  print(paste("Finished loading data after ", Sys.time() - start_time, " seconds"))
  remaining_budget <- as.integer(start_time - Sys.time() + time.budget)
  print(paste("remaining budget: ", remaining_budget, " seconds"))

  if (type == "classification") {
    train <- TaskClassif$new("benchmark_train", backend = train, target = target)
    test <- TaskClassif$new("benchmark_test", backend = test, target = target)
    if ("twoclass" %in% train$properties) {
      measure <- msr("classif.auc")
    } else {
      measure <- msr("classif.logloss")
    }
  } else if (type == "regression") {
    train <- TaskRegr$new("benchmark_train", backend = train, target = target)
    test <- TaskRegr$new("benchmark_test", backend = test, target = target)
    measure <- msr("regr.rmse")
  } else {
    stop("Task type not supported!")
  }

  model <- NULL
  training <- function() {
    model <<- AutoML(train, runtime = as.integer(remaining_budget * 0.8), measure = measure)
    model$train()
  }

  train_duration <- system.time(training())[['elapsed']]
  print(paste("Finished training model after ", difftime(Sys.time(), start_time, units = "secs"), " seconds"))

  preds <- NULL
  prediction <- function() preds <<- model$predict(test)
  predict_duration <- system.time(prediction())[['elapsed']]
  print(paste("Finished predictions after ", difftime(Sys.time(), start_time, units = "secs"), " seconds"))

  if (type == "classification") {
    sorted_colnames <- sort(colnames(preds$data$prob))
    result <- data.frame(preds$data$prob[, sorted_colnames], preds$data$response, preds$data$truth)
    colnames(result) <- c(sorted_colnames, 'predictions', 'truth')
  } else {
    result <- data.frame(preds$data$response, preds$data$truth)
    colnames(result) <- c('presdictions', 'truth')
  }

  write.csv(result, file = output_predictions_file, row.names = FALSE)

  meta_results <- data.frame(key=c("training_duration", "predict_duration"),
                             value=c(train_duration, predict_duration))
  write.csv(meta_results, file = meta_results_file, row.names = FALSE)
}

File: frameworks/mlr3automl/exec.py
import logging
import os

from amlb.benchmark import TaskConfig
from amlb.data import Dataset
from amlb.datautils import read_csv
from amlb.utils import dir_of, run_cmd

log = logging.getLogger(__name__)


def run(dataset: Dataset, config: TaskConfig):
    #TODO: use rpy2 instead? not necessary here though as the call is very simple
    log.info(f"\n**** mlr3automl (R) [{config.framework_version}] ****\n")

    here = dir_of(__file__)

    meta_results_file = os.path.join(config.output_dir, "meta_results.csv")
    run_cmd(("Rscript --vanilla -e \""
             ".libPaths('{package_directory}'); source('{script}'); "
             "run('{train}', '{test}', target.index = {target_index}, '{type}', '{output}', {cores},"
             " time.budget = {time_budget}, meta_results_file='{meta_results}', seed='{seed}', name='{name}')"
             "\"").format(
        package_directory=os.path.join(here, "lib"),
        script=os.path.join(here, 'exec.R'),
        train=dataset.train.path,
        test=dataset.test.path,
        target_index=dataset.target.index + 1,
        type=config.type,
        output=config.output_predictions_file,
        cores=config.cores,
        time_budget=config.max_runtime_seconds,
        meta_results=meta_results_file,
        seed=config.seed,
        name=config.name
    ), _live_output_=True)

    log.info("Predictions saved to %s", config.output_predictions_file)

    meta_results = read_csv(meta_results_file)
    return dict(
        training_duration=meta_result(meta_results, 'training_duration'),
        predict_duration=meta_result(meta_results, 'predict_duration')
    )


def meta_result(df, key):
    return df.loc[df['key'] == key, 'value'].squeeze()
